Create Blog “deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1”

GuopingJia opened the pull request:
Automatically generated by Netlify CMS

choukden commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,543 @@
+date: 2022-08-16T14:02:17.749Z
+author: Guoping Jia
+authorimage: /img/guoping.png
+---
```
Please add two or three tags:
tag #1: hpe-greenlake
tag #2: kubernetes
tag #3" mongodb

choukden commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,543 @@
+
+### Requirements
+
+A Kubernetes cluster can be created using either HPE GreenLake for Containers GUI, or through Infrastructure as Code (IaC) with [HPE GreenLake Terraform Provider](https://registry.terraform.io/providers/HPE/hpegl/0.2.2). By launching to the HPE GreenLake for Containers service console, you can download the **kubectl** binary, together with the *kubeconfig* file, and set it up to access the Kubern ...
```
You may want to also make reference to the blog post https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/ 

A Kubernetes cluster can be created.....with [HPE GreenLake Terraform](https://registry.terraform.io/providers/HPE/hpegl/0.2.2) as explained in the blog post [Kubernetes Cluster as Code](https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/).

choukden reviewed (changes requested):
Hi Guoping,

Thank you very much for your new great contribution to HPE Developer Community.
I have some few comments.

Dale will go through the editorial review as time permits this week or early next week.

Best regards,
Denis

GuopingJia commented:
> Hi Guoping,
> 
> Thank you very much for your new great contribution to HPE Developer Community. I have some few comments.
> 
> Dale will go through the editorial review as time permits this week or early next week.
> 
> Best regards, Denis

Thanks Denis !

The new patch has addressed your 2 comments. I will wait for Dale reviewing for more comments to address. 

Guoping

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+---
+title: Deploy Stateful MongoDB Application on Kubernetes Clusters in HPE
```
Hi there! So, HPE brand now wants us to use sentence case for titles. So the title needs to change a little here. In addition, you should either say "applications" or "a stateful MongoDB application" 

So the title could look like either of these:
Deploy stateful MongoDB applications on Kubernetes clusters in HPE GreenLake for Containers
Deploy a stateful MongoDB application on Kubernetes clusters in HPE GreenLake for Containers

I recommend the first one myself...

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+[HPE GreenLake for Containers](https://www.hpe.com/us/en/greenlake/containers.html), one of the HPE GreenLake cloud services available on the HPE GreenLake Central platform, allows customers to open the Clusters screen to create a cluster, view details about existing clusters, and launch the HPE GreenLake for Container service console. It provides an enterprise-grade container management service  ...
+
+This blog post describes the process of deploying the stateful MongoDB application on the created Kubernetes clusters in HPE GreenLake for Containers. Using the Kubernetes StatefulSet and Headless Service, together with the pre-provisioned persistent volumes, the MongoDB application can be deployed as a Replica Set that provides redundancy, fault tolerance and high availability. This MongoDB appl ...
```
I recommend changing "the stateful MongoDB" to "a stateful MongoDB"
Shouldn't "Replica Set" be "ReplicaSet"? (that's how I saw it stated in a number of places on Google :-) )
I'm not sure what you mean by "high loads in customer production environments." Should it be "and a high number of workloads in customer production environments."?

GuopingJia replied:
"ReplicaSet" is used most often in Kubernetes as an API object (at https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/), while "Replica Set" is a definition used especially in MongoDB referring to its replication architecture (at https://www.mongodb.com/docs/manual/reference/glossary/#std-term-replica-set). It does cause some confusion when I first tried it on Kubernetes. So I choose to use the "Replica Set" to comply with MongoDB manual to avoid any confusion with Kubernetes.

drensing replied:
I'm good with that. Thanks for the clarification!

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+## MongoDB Application
+
+[MongoDB](https://www.mongodb.com/) is a general-purpose, document-based NoSQL database program which is a popular alternative to traditional databases. The MongoDB data model can represent any kind of object structures which can have properties or even be nested for multiple levels. Unlike the relational databases, MongoDB can store large amounts of data without requiring a logical category or s ...
```
Change "NoSQL database program which is..." to "NoSQL database program that is..."
Similarly, in the next sentence, change "...object structures which can..." to "...object structures that can"
(There's often a lot of confusion as to when to use "that" and "which". This link might be helpful for future blog posts:
https://www.diffen.com/difference/That_vs_Which#:~:text=%22That%22%20is%20used%20to%20indicate,the%20end%20of%20the%20clause.

GuopingJia replied:
Thank you for pointing me this helpful link. Good learning :-)

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+[MongoDB](https://www.mongodb.com/) is a general-purpose, document-based NoSQL database program which is a popular alternative to traditional databases. The MongoDB data model can represent any kind of object structures which can have properties or even be nested for multiple levels. Unlike the relational databases, MongoDB can store large amounts of data without requiring a logical category or s ...
+
+MongoDB can be deployed as a standalone instance that is running as a single *mongod* server. It can also be deployed as a [Replica Set](https://www.mongodb.com/docs/manual/core/replica-set-architectures) with multiple running *mongod* instances that maintain the same data set. The Replica Set contains several data-bearing nodes. Of those data-bearing nodes, one and only one member is deemed the  ...
```
You might want to change "Replica Set" here as well to "ReplicaSet".
Question... why capitalize Primary and not secondary? If it is the name of something, you can give it an initial capital letter. So it would be "Secondary"

GuopingJia replied:
Sorry, I may miss your point. Both Primary and Secondary are capitalized that refer to the nodes in the Replica Set.

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+![](/img/mongodb-replica-set.png)
+
+The standalone MongoDB instance deployment is suitable for testing and some aspects of development. However, it’s not adequate for production use. The MongoDB Replica Set deployment ensures multiple replicas running at any given time, which provides redundancy, fault tolerance and high availability. It is recommended for production environments, such as HPE GreenLake for Containers. The blog post ...
```
"ReplicaSet" (in both instances - second sentence and last sentence)
You should probably change "The blog post" to "This blog post" in the last sentence.

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+The standalone MongoDB instance deployment is suitable for testing and some aspects of development. However, it’s not adequate for production use. The MongoDB Replica Set deployment ensures multiple replicas running at any given time, which provides redundancy, fault tolerance and high availability. It is recommended for production environments, such as HPE GreenLake for Containers. The blog post ...
+
+## Deploy MongoDB Application
```
Perhaps "Deploy a MongoDB application" would work better? (using "a" before "MongoDB" and using sentence case)

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+This blog post describes the process of deploying the stateful MongoDB application on the created Kubernetes clusters in HPE GreenLake for Containers. Using the Kubernetes StatefulSet and Headless Service, together with the pre-provisioned persistent volumes, the MongoDB application can be deployed as a Replica Set that provides redundancy, fault tolerance and high availability. This MongoDB appl ...
+
+## MongoDB Application
```
I recommend changing this to "MongoDB applications" ("s" and sentence case)

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+### Requirements
+
+A Kubernetes cluster can be created using either HPE GreenLake for Containers GUI, or through Infrastructure as Code (IaC) with [HPE GreenLake Terraform Provider](https://registry.terraform.io/providers/HPE/hpegl/0.2.2) as explained in the blog post [Kubernetes Cluster as Code](https://developer.hpe.com/blog/kubernetes-clusters-as-code-part1/). By launching to the HPE GreenLake for Containers ser ...
```
Btw, kudos to you for using "which" accurately these last couple of times :-)

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+### Set up Role-Based Access Control (RBAC)
+
+Kubernetes RBAC is a key security control to ensure that cluster users and workloads have access only to resources required to execute their roles. It is important to ensure that, when designing permissions for cluster users, the cluster administrator understands the areas where privilege escalation could occur, to reduce the risk of excessive access leading to security incidents.
```
I recommend a slight rephrasing for the first sentence:
"Kubernetes RBAC is a key security control **used** to ensure..."
You don't really need the comma after "occur", btw

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+### Create the StatefuleSet Deployment
+
+To deploy MongoDB as a Replica Set with multiple pods, a Kubernetes StatefulSet deployment will be required. The data persistence setup can be done with a *VolumeClaimTemplate* in the StatefulSet deployment.
```
"ReplicaSet" https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+          storage: 1Gi
+```
+
+### Set Up the Headless Service
````
"Set **up** the Headless Service" (from what I can see, Headless Service is a proper noun, hence the initial capitalization). But you still need to change "up"

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+statefulset.apps/pce-mongodb-replica created
+```
+
+MongoDB application is deployed as Replica Set to the namespace '*pce-mongodb*' using the service account '*pce-mongo-account*'. Running the following command, you should see three specified replicas pods are all in **running** status: 
````
ReplicaSet

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+statefulset.apps/pce-mongodb-replica   3/3     30s
+```
+
+## Set up MongoDB Replication Host
````
Is MongoDB Replication Host a proper noun? Or should it be MongoDB replication host?

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+## Set up MongoDB Replication Host
+
+After MongoDB application is deployed with all the replica pods in running states, you need to set up MongoDB replication. 
```
After **the** MongoDB application...

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+rs0:PRIMARY>
+```
+
+#### 4. Add the second MongoDB pod *pce-mongodb-replica-1* to the replication configuration:
````
"second" or "Secondary"? or "secondary"?

GuopingJia replied:
The MongoDB app has been deployed with 3 pods, 'pce-mongodb-replica-[0,1,2]'. The first pod 'pce-mongodb-replica-0' has been added as the Primary server (step 3). Here, the *second* pod, i.e., 'pce-mongodb-replica-1', needs to be added to the configuration. Similarly, the next step, step 5, is to add the *third* pod 'pce-mongodb-replica-2' to the configuration. Since the MongoDB Replica Set has one and only one Primary node, the *second* and the *third* pod will be added as the Secondary nodes.

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+rs0:PRIMARY>
+```
+
+The **members** section of the status output shows three replicas. The pod *pce-mongodb-replica-0* is listed as the **PRIMARY** replica, while the other two pods, *pce-mongodb-replica-1* & *pce-mongodb-replica-2*, are listed as the **SECONDARY** replicas.
````
Should PRIMARY and SECONDARY be all caps? Is that how they are actually named or shown on a screen? Try to keep consistent throughout the blog post.

GuopingJia replied:
Good catch. I will change them to Primary and Secondary to be more consistent.

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
````diff
@@ -0,0 +1,547 @@
+$ kubectl scale statefulset <name> -n <namespace> --replicas=<number of replicas>
+```
+
+Then you follow up the *step 1* in previous section to connect to the first MongoDB pod and repeat the *step 5* to add the new replica pod.
````
Recommend rephasing this to "Then, follow up with *Step 1* from the previous section to connect to the first MongoDB pod and repeat *Step 5* to add the new replica pod."

drensing commented on content/blog/deploy-stateful-mongodb-application-on-kubernetes-clusters-in-hpe-greenlake-for-containers-1.md:
```diff
@@ -0,0 +1,547 @@
+
+## Summary
+
+This blog post describes the detailed process to deploy and set up a stateful MongoDB application as a Replica Set deployment on Kubernetes clusters in HPE GreenLake for Containers. Kubernetes provides the ability to persist the state, deploy the stateful application, and manage and scale those applications with state. The production deployment of the MongoDB application provides redundancy, faul ...
```
Slight rephrasing of the first sentence:
This blog post describes the detailed process used to deploy..."
ReplicaSet?
Modify second to last sentence ending to "...fault tolerance, high availability, and scalability."
Modify very last sentence to make that "high loads" section more clear as you did earlier.

drensing reviewed (commented):
Thank you again for such a great detailed tutorial! Your posts keep getting better and better. Really minor modifications. The only reason you see a number of them is because of the same mistake made a couple of times. Great job!

GuopingJia commented:
> Thank you again for such a great detailed tutorial! Your posts keep getting better and better. Really minor modifications. The only reason you see a number of them is because of the same mistake made a couple of times. Great job!

Thanks Dale for those good comments to my draft blog ! I highly appreciate your good suggestion and the helpful link you provided. 

The new patch has addressed your comments. But let me know if there is still anything you want me to change.  

Guoping

choukden reviewed (changes requested):
Hi Jia,

I believe your blog post is ready for publication.
Could you please confirm you have incorporated all suggestions for edits from Dale and the post is now ready for publication?

Is so, I will publish the post in the HPE Dev Community portal.

Thanks,
Denis

GuopingJia commented:
> Hi Jia,
> 
> I believe your blog post is ready for publication. Could you please confirm you have incorporated all suggestions for edits from Dale and the post is now ready for publication?
> 
> Is so, I will publish the post in the HPE Dev Community portal.
> 
> Thanks, Denis

Hi, Denis, 

Thank you for asking this. 

Yes, I can confirm last 2 patches have been addressed both your and Dale's comments/suggestions. 
The draft is ready for your approval / publication. 

Guoping
